### DISCUSSION TITLE
Astro Integration / Adapter to support Electron and Browser Extensions

### DISCUSSION THREAD

**Discussion by a-y-u-s-h** [withastro/roadmap#807] [Proposal]
# Summary

This is a feature request for addition to two adapters / integrations `@astro/electron` and `@astro/extensions` - former takes in the output (and some electron specific files) and converts it into an electron based application, and the latter takes in the ouput (and some browser extensions specific files) and lets you output a browser extension, like a Chrome / Firefox extension based on other config parameters.

# Background & Motivation

Currently Astro is a Web Framework, with pretty decent adapters, integrations that take the output and have it work in different environments. Some integrations take in additional files to modify the output (Like Tailwind takes in Tailwind Config). That led me to think that adapters and integrations to build / output desktop apps with Electron (say `@astro/electron`), or an adapter to output chrome or firefox or browser extensions in general (say `@astro/extensions`) are a possibility. I do not know how to do it, so I am proposing it here as a feature request. 

## Why this RFC is important?

1. It opens up lots of use cases and upgrades Astro from being a Web only framework to support platforms that Javascript world has touched on in recent years - Desktop, Browser Extensions. 

2. 1st order consequence of having these integrations are that you're able to tap into additional capabilities provided by these platforms by a single plugin, because the JS world already supports these platforms. 

3. 2nd order consequence of having these adapters / integration could be that eventually this may lead to a foundation for support of the devices where the most number of users live on - the mobile phones with another `@astro/mobile` package and Astro can graduate from a Web-only to a general framework (through plugin system) while keeping everything decoupled and opt-in. 

4. Users on discord have expressed their interest in using Astro's simplicity to produce chrome extensions and electron apps. There are some examples of people using [Astro with Tauri](https://github.com/JonasKruckenberg/tauri-astro-template) but none yet for Electron, and besides I think if these were officially supported plugins, they would integrate better with the framework (an `@astro/tauri` could be a thing as well perhaps but since Electron seems to have a larger developer base because it's javascript based so I'm inclined to suggest its integration first personally).

# Goals

A **concise, bulleted-list** outlining the intended goals of this RFC. 

- Introduce @astro/electron and @astro/extensions adapters for seamless conversion to Electron-based apps and browser extensions.
- Expand AstroJS support beyond web-centric frameworks for enhanced platform compatibility.
- Ensure user-friendly integration, prioritizing a smooth developer experience.
- Enable developers to tap into specific Electron and extension capabilities effortlessly.
- Maintain a decoupled and opt-in approach for flexible feature adoption.
- Establish a foundation for potential future integrations, such as mobile device support (@astro/mobile).
- Seek community validation and feedback to meet developer needs collaboratively.

# Example

```js
import electron from "@astro/electron"

export default {
 ...
  adapter: electron({
    ...
  })
  ...
}

```

```js
import extensions from "@astro/extensions"

export default {
 ...
  adapter: extensions({
    target: "chrome",
    ...
  })
  ...
}

```

**Comment by lilnasy** [comment]
Is there something preventing this from being a third party integration?

**Reply by a-y-u-s-h** [reply]
I don't think so (I don't really know the criteria, or how an unofficial plugin becomes official). Personally, I haven't been able to create these due to lack of updated knowledge about those platforms (electron especially) and bandwidth yet so putting the idea out there for feedback and comments.

**Reply by NickChasanis** [reply]
I think this is a legit ask. 
Would like to see more details

**Reply by delucis** [reply]
Not quite what was asked, but might still be a handy reference. This is a template for building Tauri apps with Astro: https://github.com/JonasKruckenberg/tauri-astro-template

**Reply by a-y-u-s-h** [reply]
> Not quite what was asked, but might still be a handy reference. This is a template for building Tauri apps with Astro: https://github.com/JonasKruckenberg/tauri-astro-template

I mentioned that project in 4th point of "Why this RFC is important?" as well. I guess at the time I wanted electron support because that way it'll be an easy transition for developers who wish to work only within JS/TS ecosystem. I still think it would be nice to have for many reasons, one of them being that electron has a large community around it.
